session_test.go

```diff
@@ -67,14 +68,152 @@ func (p *pipeConn) Close() error {
 	return p.writer.Close()
 }
 
-func testConn() (io.ReadWriteCloser, io.ReadWriteCloser) {
+func testConnPipe(testing.TB) (io.ReadWriteCloser, io.ReadWriteCloser) {
 	read1, write1 := io.Pipe()
 	read2, write2 := io.Pipe()
 	conn1 := &pipeConn{reader: read1, writer: write2}
 	conn2 := &pipeConn{reader: read2, writer: write1}
 	return conn1, conn2
 }
 
+func testConnTCP(t testing.TB) (io.ReadWriteCloser, io.ReadWriteCloser) {
+	l, err := net.ListenTCP("tcp", nil)
+	if err != nil {
+		t.Fatalf("error creating listener: %v", err)
+	}
+	t.Cleanup(func() { _ = l.Close() })
+
+	network := l.Addr().Network()
+	addr := l.Addr().String()
+
+	var server net.Conn
+	errCh := make(chan error, 1)
+	go func() {
+		defer close(errCh)
+		var err error
+		server, err = l.Accept()
+		if err != nil {
+			errCh <- err
+			return
+		}
+	}()
+
+	t.Logf("Connecting to %s: %s", network, addr)
+	client, err := net.DialTimeout(network, addr, 10*time.Second)
+	if err != nil {
+		t.Fatalf("error dialing tls listener: %v", err)
+	}
+	t.Cleanup(func() { _ = client.Close() })
+
+	if err := <-errCh; err != nil {
+		t.Fatalf("error creating tls server: %v", err)
+	}
+	t.Cleanup(func() { _ = server.Close() })
+
+	return client, server
+}
+
+func testConnTLS(t testing.TB) (io.ReadWriteCloser, io.ReadWriteCloser) {
+	cert, err := tls.LoadX509KeyPair("testdata/cert.pem", "testdata/key.pem")
+	if err != nil {
+		t.Fatalf("error loading certificate: %v", err)
+	}
+
+	l, err := net.ListenTCP("tcp", nil)
+	if err != nil {
+		t.Fatalf("error creating listener: %v", err)
+	}
+	t.Cleanup(func() { _ = l.Close() })
+
+	var srv net.Conn
```

Comment:
This feels slightly racy. Probably better to send this `net.Conn` back out of the goroutine using another channel?

Reply:
Same as above, errCh synchronizes access to srv... although switching naming schemes is a bit annoying. I'll fix that up.